src/erc721.rs

```diff
@@ -0,0 +1,367 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+use sp_runtime::traits::Hash as HashT;
+use support::{
+    decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchError,
+    dispatch::DispatchResult, ensure, traits::Randomness,
+};
+use system::ensure_signed;
+
+// Encoding library
+use codec::Encode;
+
+pub trait Trait: system::Trait + balances::Trait + contracts::Trait {
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    // Something that provides randomness generation in the runtime.
+    type Randomness: Randomness<Self::Hash>;
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        // No owner set for token
+        TokenNoOwner,
+
+        // Got an overflow after adding
+        Overflow,
+
+        // Got an underflow after subing
+        Underflow,
+    }
+}
+
+decl_event!(
+    pub enum Event<T> 
+        where
+        <T as system::Trait>::AccountId,
+        <T as system::Trait>::Hash {
+
+        // Token transfer event
+        Transfer(Option<AccountId>, Option<AccountId>, Hash),
+
+        // One token approved to an account
+        Approval(AccountId, AccountId, Hash),
+
+        // All tokens owned are approved to other account
+        ApprovalForAll(AccountId, AccountId, bool),
+    }
+);
+
+decl_storage! {
+    trait Store for Module<T: Trait> as ERC721 {
+        // Token count owned by an account
+        pub OwnedTokensCount get(balance_of): map T::AccountId => u64;
+
+        // Token owner's account id
+        pub TokenOwner get(owner_of): map T::Hash => Option<T::AccountId>;
+
+        // Token delegate's account id
+        pub TokenApprovals get(get_approved): map T::Hash => Option<T::AccountId>;
+
+        // Account deledate's account id
+        pub OperatorApprovals get(is_approved_for_all): map (T::AccountId, T::AccountId) => bool;
+
+        // Total count of minted token
+        pub TotalSupply get(total_supply): u64;
```

Comment:
In Solidity implementation, every uint is uint256 (256 bits).  If we want to follow the standard, we need more bits.

I just looked it up in Rust the native support is up to `u128` https://doc.rust-lang.org/std/primitive.u128.html 

Do you think we should use some big number library instead of `u64`?

Reply:
I will check if wasm can support it well. And maybe big number already in substrate framework. I am wondering now if it is a better idea to leave the type in Trait definition. Then ERC721 user can decide the type during building the runtime.

Reply:
@juniuszhou I like your proposal, let's keep it as a type in the Trait.